### PR TITLE
[KDESKTOP-1101] Two optimisations of UC time in utility functions

### PR DESCRIPTION
**Pull request by herve-er**
## CommonUtility::maxPathLength():
- [x] Done

The CommonUtility::maxPathLength() function on Windows currently accounts for 8.1% of the UC time durring step 1, which is disproportionate for a utility function. To drastically reduce this UC time, I propose to retrieve the max path length once at startup and store it in a static variable. This avoids querying the register key at each call, significantly improving performance.

The risk of someone changing the registry key to a lower value during runtime is highly improbable, and the consequences do not justify the performance degradation. Any issues caused by such changes can be resolved with a system reboot.
Not unit testable.
### Results (UC time):
     - Before: 8.1% (+- 1m)
     - After: <0.01% (+- 0.025s) 🚀 
 
-----------------------------

## IoHelper::checkIfPathExists:
- [x] Done
The IoHelper::checkIfPathExists function on Windows currently accounts for 15.6% of the UC time during step 2, which seems high for an IoHelper function. After investigation, it appears that this function currently relies on IoHelper::getItemType, which gathers a lot of information on the file that we do not really need here. I have modified it with the goal of reducing UC consumption. It will now only gather the information needed and if it is needed.
Already unit tested.
### Results (UC time):
     - Before: 15.6% (+- 35s)
     - After: 11.4% (+- 25s) ⏩

### CODE REVIEW

File: src/libcommon/utility/utility.cpp [hunk 1]
```diff
@@ -75,6 +75,9 @@ namespace KDC {
 
 const int CommonUtility::logsPurgeRate = 7;               // days
 const int CommonUtility::logMaxSize = 500 * 1024 * 1024;  // MB
+#if defined(_WIN32)
+size_t CommonUtility::_maxPathWin = 0;
```
Comment (luc-guyot-infomaniak) [inline src/libcommon/utility/utility.cpp:79]: You can define a static variable within the function `maxPathLength` itself. This way we don't create a variable the rest of the application should not be aware of (encapsulation).

File: src/libcommon/utility/utility.cpp [hunk 2]
```diff
@@ -619,13 +622,16 @@ QString CommonUtility::getRelativePathFromHome(const QString &dirPath) {
 
 size_t CommonUtility::maxPathLength() {
 #if defined(_WIN32)
-    Poco::Util::WinRegistryKey key(R"(HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\FileSystem)", true);
-    bool exist = key.exists("LongPathsEnabled");
-    if (exist && key.getInt("LongPathsEnabled")) {
-        return MAX_PATH_LENGTH_WIN_LONG;
-    } else {
-        return MAX_PATH_LENGTH_WIN_SHORT;
+    if (_maxPathWin == 0) {
```
Comment (luc-guyot-infomaniak) [inline src/libcommon/utility/utility.cpp:625]:
```suggestion
    static size_t maxPathWin = 0;
    if (maxPathWin == 0) {
```
The initialization will be done on the first call to the function only.

File: src/libcommon/utility/utility.cpp [hunk 3]
```diff
@@ -636,15 +642,9 @@ size_t CommonUtility::maxPathLength() {
...
-    if (exist && key.getInt("LongPathsEnabled")) {
-        return MAX_PATH_LENGTH_WIN_LONG;
-    } else {
-        return MAX_PATH_LENGTH_WIN_SHORT -
-               12;  // For folders in short path mode, it is MAX_PATH - 12
-                    // (https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation?tabs=registry)
-    }
+    // For folders in short path mode, it is MAX_PATH - 12
+    // (https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation?tabs=registry)
+    return maxPathLength() == MAX_PATH_LENGTH_WIN_LONG ? MAX_PATH_LENGTH_WIN_LONG : MAX_PATH_LENGTH_WIN_SHORT - 12;
```
Comment (luc-guyot-infomaniak) [inline src/libcommon/utility/utility.cpp:647]:
```suggestion
    return (maxPathLength() == MAX_PATH_LENGTH_WIN_LONG) ? MAX_PATH_LENGTH_WIN_LONG : MAX_PATH_LENGTH_WIN_SHORT - 12;
```

File: test/libcommonserver/io/testcheckifpathexists.cpp [hunk 4]
```diff
@@ -96,10 +96,45 @@ void TestIo::testCheckIfPathExistsSimpleCases() {
         CPPUNIT_ASSERT(exists);
         CPPUNIT_ASSERT(ioError == IoErrorSuccess);
     }
+
+    // A regular file withour read/write permission
```
Comment (luc-guyot-infomaniak) [inline test/libcommonserver/io/testcheckifpathexists.cpp:100]:
```suggestion
    // A regular file without read/write permission
```

File: test/libcommonserver/io/testcheckifpathexists.cpp [hunk 5]
```diff
@@ -96,10 +96,45 @@ void TestIo::testCheckIfPathExistsSimpleCases() {
...
+    {
+        LocalTemporaryDirectory temporaryDirectory("TestIo");
+        const SyncPath path = temporaryDirectory.path() / "test.txt";
+        std::ofstream ofs(path);
+        ofs << "Some content.\n";
```
Comment (luc-guyot-infomaniak) [inline test/libcommonserver/io/testcheckifpathexists.cpp:105]: Is it useful?

File: test/libcommonserver/io/testcheckifpathexists.cpp [hunk 6]
```diff
@@ -96,10 +96,45 @@ void TestIo::testCheckIfPathExistsSimpleCases() {
         CPPUNIT_ASSERT(exists);
         CPPUNIT_ASSERT(ioError == IoErrorSuccess);
     }
+
+    // A regular file withour read/write permission
+    {
+        LocalTemporaryDirectory temporaryDirectory("TestIo");
+        const SyncPath path = temporaryDirectory.path() / "test.txt";
+        std::ofstream ofs(path);
```
Comment (luc-guyot-infomaniak) [inline test/libcommonserver/io/testcheckifpathexists.cpp:104]:
```suggestion
        {std::ofstream(path);}
```
The destructor of `std::ofstream` will close the file.

File: test/libcommonserver/io/testcheckifpathexists.cpp [hunk 7]
```diff
@@ -96,10 +96,45 @@ void TestIo::testCheckIfPathExistsSimpleCases() {
...
+        ofs << "Some content.\n";
+        ofs.close();
+
+        IoError ioError = IoErrorUnknown;
+        bool setRightResults = IoHelper::setRights(path, false, false, false, ioError) && ioError == IoErrorSuccess;
```
Comment (luc-guyot-infomaniak) [inline test/libcommonserver/io/testcheckifpathexists.cpp:109]:
```suggestion
        const bool setRightResults = IoHelper::setRights(path, false, false, false, ioError) && ioError == IoErrorSuccess;
```

File: test/libcommonserver/io/testcheckifpathexists.cpp [hunk 8]
```diff
@@ -96,10 +96,45 @@ void TestIo::testCheckIfPathExistsSimpleCases() {
...
+        if (!setRightResults) {
+            IoHelper::setRights(path, true, true, true, ioError);
+            CPPUNIT_FAIL("Failed to set rights on the file");
+        }
+        bool exists = false;
+        bool checkIfPathExistsResult = _testObj->checkIfPathExists(path, exists, ioError);
```
Comment (luc-guyot-infomaniak) [inline test/libcommonserver/io/testcheckifpathexists.cpp:115]:
```suggestion
        const bool checkIfPathExistsResult = _testObj->checkIfPathExists(path, exists, ioError);
```

File: test/libcommonserver/io/testcheckifpathexists.cpp [hunk 9]
```diff
@@ -96,10 +96,45 @@ void TestIo::testCheckIfPathExistsSimpleCases() {
...
+
+        CPPUNIT_ASSERT(checkIfPathExistsResult);
+        CPPUNIT_ASSERT(exists);
+        CPPUNIT_ASSERT(ioError == IoErrorSuccess);
+    }
+
+    // Checking existence of a subdirectory inside a directory that have been deleted and replaced with a file with the same name
+    // ex: conversion of a bundle into a single file (macOS)
+    {
+        const SyncPath path = _localTestDirPath / "test_pictures/picture-1.jpg/A";
```
Comment (luc-guyot-infomaniak) [inline test/libcommonserver/io/testcheckifpathexists.cpp:126]:
```suggestion
        const SyncPath path = _localTestDirPath / "test_pictures" / "picture-1.jpg" / "A";
```

File: test/libcommonserver/io/testcheckifpathexists.cpp [hunk 10]
```diff
@@ -333,10 +338,38 @@ void TestIo::testCheckIfPathExistsWithSameNodeIdSimpleCases() {
         CPPUNIT_ASSERT(ioError == IoErrorSuccess);
     }
 
+    // A regular file without read/write permission
+    {
+        LocalTemporaryDirectory temporaryDirectory("TestIo");
+        const SyncPath path = temporaryDirectory.path() / "test.txt";
+        std::ofstream ofs(path);
```
Comment (luc-guyot-infomaniak) [inline test/libcommonserver/io/testcheckifpathexists.cpp:345]: Same remarks as above.

File: test/libcommonserver/io/testcheckifpathexists.cpp [hunk 11]
```diff
@@ -333,10 +338,38 @@ void TestIo::testCheckIfPathExistsWithSameNodeIdSimpleCases() {
...
+        ofs.close();
+
+        IoError ioError = IoErrorUnknown;
+        bool existsWithSameId = false;
+        NodeId otherNodeId;
+        NodeId nodeId;
+
+        CPPUNIT_ASSERT(_testObj->getNodeId(path, nodeId));
+
+        bool setRightResults = IoHelper::setRights(path, false, false, false, ioError) && ioError == IoErrorSuccess;
```
Comment (luc-guyot-infomaniak) [inline test/libcommonserver/io/testcheckifpathexists.cpp:356]:
```suggestion
        const bool setRightResults = IoHelper::setRights(path, false, false, false, ioError) && ioError == IoErrorSuccess;
```

File: test/libcommonserver/io/testcheckifpathexists.cpp [hunk 12]
```diff
@@ -378,106 +411,67 @@ void TestIo::testCheckIfPathExistsWithSameNodeIdSimpleCases() {
...
 
-    // Reading a regular symbolic link that is removed right after `filesystem::is_simlink` was called.
+    // A Windows junction on a regular target file.
     {
-        const SyncPath targetPath = _localTestDirPath / "test_pictures";
-        const LocalTemporaryDirectory temporaryDirectory;
-        const SyncPath path = temporaryDirectory.path() / "regular_dir_symbolic_link";
-        std::filesystem::create_symlink(targetPath, path);
+        const LocalTemporaryDirectory temporaryDirectory("TestIo");
+        const SyncPath targetPath = _localTestDirPath / "test_pictures/picture-1.jpg";
```
Comment (luc-guyot-infomaniak) [inline test/libcommonserver/io/testcheckifpathexists.cpp:459]:
```suggestion
        const SyncPath targetPath = _localTestDirPath / "test_pictures" / "picture-1.jpg";
```

File: test/libcommonserver/io/testcheckifpathexists.cpp [hunk 13]
```diff
@@ -378,106 +411,67 @@ void TestIo::testCheckIfPathExistsWithSameNodeIdSimpleCases() {
...
+        const SyncPath path = temporaryDirectory.path() / "dir_junction";
 
-        _testObj->setReadSymlinkFunction([](const SyncPath &path, std::error_code &ec) -> SyncPath {
-            std::filesystem::remove(path);
-            return std::filesystem::read_symlink(path, ec);
-        });
+        IoError ioError = IoErrorUnknown;
+        CPPUNIT_ASSERT(_testObj->createJunctionFromPath(targetPath, path, ioError));
```
Comment (luc-guyot-infomaniak) [inline test/libcommonserver/io/testcheckifpathexists.cpp:463]: It's surprising that it works. Good to know anyway.

**Review by luc-guyot-infomaniak — COMMENTED** [review]
Comments about function's static variable, constness and junction targetting a file.

File: test/libcommonserver/io/testcheckifpathexists.cpp (hunk 5 above)
Comment (herve-er) [reply test/libcommonserver/io/testcheckifpathexists.cpp]: [4f58023](https://github.com/Infomaniak/desktop-kDrive/pull/239/commits/4f58023ff3bd5aff09ada2606e377a3d08b28a9b)

File: test/libcommonserver/io/testcheckifpathexists.cpp (hunk 6 above)
Comment (herve-er) [reply test/libcommonserver/io/testcheckifpathexists.cpp]: [4f58023](https://github.com/Infomaniak/desktop-kDrive/pull/239/commits/4f58023ff3bd5aff09ada2606e377a3d08b28a9b)

File: test/libcommonserver/io/testcheckifpathexists.cpp (hunk 13 above)
Comment (herve-er) [reply test/libcommonserver/io/testcheckifpathexists.cpp:458]: Not sure about what you are talking about here?

(I just change the name of the junction if it was what you where talking about 👍🏾 @herve-er
[More explicit name for test junction.](https://github.com/Infomaniak/desktop-kDrive/pull/239/commits/a3f6111fcb7cd151fd00d1f7e80cbc9b9709b437))

File: src/libcommon/utility/utility.cpp (hunk 1 above)
Comment (herve-er) [reply src/libcommon/utility/utility.cpp]: @herve-er
[define maxPathWin in maxPathLength()](https://github.com/Infomaniak/desktop-kDrive/pull/239/commits/b2b371c7f7e5c0cd809c1c438adfd2c98e035bc5)

File: test/libcommonserver/io/testcheckifpathexists.cpp (hunk 13 above)
Comment (luc-guyot-infomaniak) [reply test/libcommonserver/io/testcheckifpathexists.cpp:458]: I was referring to the outcome of the test: we can create successfully a junction whose target is a file, as opposed to a directory. It surprised me when I re-discovered it.

File: test/libcommonserver/io/testcheckifpathexists.cpp [hunk 14]
```diff
@@ -96,10 +96,42 @@ void TestIo::testCheckIfPathExistsSimpleCases() {
...
+        bool exists = false;
+        const bool checkIfPathExistsResult = _testObj->checkIfPathExists(path, exists, ioError);
+        IoHelper::setRights(path, true, true, true, ioError);
+
+        CPPUNIT_ASSERT(checkIfPathExistsResult);
+        CPPUNIT_ASSERT(exists);
+        CPPUNIT_ASSERT(ioError == IoError::Success);
+    }
+
+    // Checking existence of a subdirectory inside a directory that have been deleted and replaced with a file with the same name
```
Comment (luc-guyot-infomaniak) [inline test/libcommonserver/io/testcheckifpathexists.cpp:120]:
```suggestion
    // Checking existence of a subdirectory inside a directory that has been deleted and replaced with a file with the same name.
```

File: test/libcommonserver/io/testcheckifpathexists.cpp [hunk 15]
```diff
@@ -96,10 +96,42 @@ void TestIo::testCheckIfPathExistsSimpleCases() {
...
+    // ex: conversion of a bundle into a single file (macOS)
```
Comment (luc-guyot-infomaniak) [inline test/libcommonserver/io/testcheckifpathexists.cpp:121]:
```suggestion
    // Example: the conversion of a bundle into a single file (macOS).
```

**Review by luc-guyot-infomaniak — COMMENTED** [review]
Minor comments on comments in unit tests

**Review by luc-guyot-infomaniak — APPROVED** [review]

**Pull request merged**